zerver/tests/test_auth_backends.py

```diff
@@ -1159,16 +1160,53 @@ def test_social_auth_saml_multiple_idps_configured(self) -> None:
         should lead to misconfiguration page.
         """
 
-        with self.settings(SOCIAL_AUTH_SAML_ENABLED_IDPS={"test_idp1": {}, "test_idp2": {}}):
-            # We don't need to put full idp configurations in the mock settings above
-            # to trigger the error.
-            with mock.patch("zerver.views.auth.logging.error") as mock_error:
-                result = self.client_get("/accounts/login/social/saml")
-                self.assertEqual(result.status_code, 302)
-                self.assertEqual(result.url, '/config-error/saml')
-                mock_error.assert_called_once_with(
-                    "SAML misconfigured - you have specified multiple IdPs. Only one IdP is supported."
-                )
+        # Setup a new SOCIAL_AUTH_SAML_ENABLED_IDPS dict with two idps.
+        # We deepcopy() dictionaries around for the sake of brevity,
+        # to avoid having to spell them out explicitly here.
+        # The second idp's configuration is a copy of the first one,
+        # with name test_idp2 and altered url.
+        idps_dict = copy.deepcopy(settings.SOCIAL_AUTH_SAML_ENABLED_IDPS)
+        idps_dict['test_idp2'] = copy.deepcopy(idps_dict['test_idp'])
+        idps_dict['test_idp2']['url'] = 'https://idp2.example.com/idp/profile/SAML2/Redirect/SSO'
+        idps_dict['test_idp2']['display_name'] = 'Second Test IdP'
+
+        # Run tests with multiple idps configured:
+        with self.settings(SOCIAL_AUTH_SAML_ENABLED_IDPS=idps_dict):
+            # Go to the login page and check that buttons to log in show up for both IdPs:
+            result = self.client_get('/accounts/login/')
+            self.assert_in_success_response(["Log in with Test IdP"], result)
+            self.assert_in_success_response(["/accounts/login/social/saml/test_idp"], result)
+            self.assert_in_success_response(["Log in with Second Test IdP"], result)
+            self.assert_in_success_response(["/accounts/login/social/saml/test_idp2"], result)
+
+            # Try succesful authentication with the regular idp from all previous tests:
+            self.test_social_auth_success()
+
+            # Now test with the second idp:
+            self.LOGIN_URL = "/accounts/login/social/saml/test_idp2"
+            self.SIGNUP_URL = "/accounts/register/social/saml/test_idp2"
+            self.AUTHORIZATION_URL = idps_dict['test_idp2']['url']
+            self.test_social_auth_success()
+
+        self.LOGIN_URL = "/accounts/login/social/saml/test_idp"
+        self.SIGNUP_URL = "/accounts/register/social/saml/test_idp"
+        self.AUTHORIZATION_URL = idps_dict['test_idp']['url']
```

Comment:
So this monkey-batching of these settings is a bit dangerous, in that it can leak outside the scope we intend it to.  A `try/finally` block would be ideal, and I think it'd also be best to save these values in an `orig_LOGIN_URL` variable so that this code always restores the original values even if we change them in the future.

Reply:
Changed:
```
            try:
                original_LOGIN_URL = self.LOGIN_URL
                original_SIGNUP_URL = self.SIGNUP_URL
                original_AUTHORIZATION_URL = self.AUTHORIZATION_URL
                self.LOGIN_URL = "/accounts/login/social/saml/test_idp2"
                self.SIGNUP_URL = "/accounts/register/social/saml/test_idp2"
                self.AUTHORIZATION_URL = idps_dict['test_idp2']['url']
                self.test_social_auth_success()
            finally:
                # Restore original values at the end, regardless of what happens
                # in the block above, to avoid affecting other tests in unpredictable
                # ways.
                self.LOGIN_URL = original_LOGIN_URL
                self.SIGNUP_URL = original_SIGNUP_URL
                self.AUTHORIZATION_URL = original_AUTHORIZATION_URL
```